designs/builtins-design.md

```diff
@@ -0,0 +1,201 @@
+
+#### Import Handling
+We plan to place the import handling stage right after parsing before doing the type checking. We will use the information in AST, such as `imports` and function calls, to preprocess the input source code. For example, if we get the following source in its AST format, we will be able to recognize the import statement and the corresponding function calls depending on that import:
+
+```python
+from math import gcd, lcm
+print(gcd(12, 16), lcm(12, 16))
+```
+
+Then we rewrite the AST to include our library implementation, essentially replacing the original source file with the updated one below:
+
+```diff
+-from math import gcd, lcm
++def gcd(a: int, b: int) -> int:
++    if(b==0):
++        return a
++    else:
++        return gcd(b,a%b)
+
++def lcm(x: int, y: int) -> int:
++   return (x*y)//gcd(x,y)
++
+print(gcd(12, 16), lcm(12, 16))
+```
+
+After that the compilation procedure would continue as if the new AST were the original program. 
+
+The built-in library will live inside `stdlib/*.py` or `stdlib/*.wat` and will be copied to `build/stdlib/*` so the frontend can fetch the files. We do the AST updates using Python files fetched through HTTP GET at compile time, importing `math` module would mean a request to `/stdlib/math.py`. 
+
+In our first iteration, we'll focus on getting the basic import functionality work, meaning:
+- no circular imports allowed (catching this will be kind of annoying, so we won't implement it yet)
+- during AST construction, add `FunDef`s corresponding to all ImportedModules' functions to `Program.funs`; this will make type-checking seamless
+- add individual functions imported from modules into the AST the same way
+- for built-in functions available globally without imports, e.g. `len, max` we inject the function implementation when we see the corresponding call to it
+
+#### AST
+- add a new `ImportedModule` type:
+```typescript
+export type ImportedModule = {
+    name: string,
+    importedAs: string
+}
+```
+- add a new field `imports: Array<ImportedModule>` to `Program`
+- add new Statement type called `import`, will be eliminated by Import Handling pass before lowering to IR
+- add new Literal kind `...` (but don't add its type (?), because casting to an ellipsis doesn't make sense)
+- add new Literal kind `float: number`
+
+#### IR
+- we won't make changes to IR, at least not in the next few weeks.
+
+#### Built-in libraries
+
+- print() will get changed pretty extensively
+    - print will no longer be unary -- any arity is valid!
+        - nullary: prints new line
+        - unary: prints the argument
+        - n-ary: prints all arguments, separated by spaces
+
+## New functions, types, and/or files
+- `stdlib/math.py`
+    - factorial(n : int) -> int
+    - gcd(a : int, b : int) -> int
+    - lcm(a : int, b : int) -> int
+- `float`, as a type
+- various flavors of `import` statement
```

Comment:
Is there a “most basic” one you can do for week 1, maybe `from x import y, z` since it doesn't involve a new meaning for the `.` operator?